src/solvers/refinement/string_constraint_generator.cpp

```diff
@@ -0,0 +1,2147 @@
+  equal_exprt res_eq(s2[idx2], res[plus_exprt(idx2, s1.length())]);
+  string_constraintt a2(idx2, s2.length(), res_eq);
+  axioms.push_back(a2);
+
+  return res;
+}
+
+
+string_exprt string_constraint_generatort::add_axioms_for_concat
+(const function_application_exprt &f)
+{
+  const function_application_exprt::argumentst &args=f.arguments();
+  assert(args.size()==2);
+
+  string_exprt s1=add_axioms_for_string_expr(args[0]);
+  string_exprt s2=add_axioms_for_string_expr(args[1]);
+
+  return add_axioms_for_concat(s1, s2);
+}
+
+
+string_exprt string_constraint_generatort::add_axioms_for_copy
+(const function_application_exprt &f)
+{
+  string_exprt s1=add_axioms_for_string_expr(args(f, 1)[0]);
+  string_exprt res(get_char_type());
+  axioms.push_back(res.same_length(s1));
+  symbol_exprt idx=fresh_univ_index("QA_index_copy");
+  string_constraintt a1(idx, s1.length(), equal_exprt(s1[idx], res[idx]));
+  axioms.push_back(a1);
+  return res;
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_set_length
+(const function_application_exprt &f)
+{
+  string_exprt s1=add_axioms_for_string_expr(args(f, 2)[0]);
+  exprt k=args(f, 2)[1];
+  string_exprt res(get_char_type());
+
+  // |s|=k
+  // && forall i<k. (i<k ==> s[i]=s1[i]) && (i >= k ==> s[i]=0)
+
+  axioms.push_back(res.has_length(k));
+  symbol_exprt idx=fresh_univ_index("QA_index_set_length");
+  string_constraintt a1
+    (idx, k, and_exprt
+      (implies_exprt(s1.strictly_longer(idx), equal_exprt(s1[idx], res[idx])),
+       implies_exprt(s1.shorter(idx), equal_exprt(s1[idx], constant_char(0)))));
+  axioms.push_back(a1);
+
+  return res;
+}
+
+
+string_exprt string_constraint_generatort::add_axioms_for_java_char_array
+(const exprt & char_array)
+{
+  string_exprt res(get_char_type());
+  exprt arr=to_address_of_expr(char_array).object();
+  exprt len=member_exprt(arr, "length", res.length().type());
+  exprt cont=member_exprt(arr, "data", res.content().type());
+  res.op0()=len;
+  res.op1()=cont;
+  return res;
+}
+
+
+string_exprt string_constraint_generatort::add_axioms_for_value_of
+(const function_application_exprt &f)
+{
+  const function_application_exprt::argumentst &args=f.arguments();
+  if(args.size()==3)
+    {
+      string_exprt res(get_char_type());
+      exprt char_array=args[0];
+      exprt offset=args[1];
+      exprt count=args[2];
+      string_exprt str=add_axioms_for_java_char_array(char_array);
+      axioms.push_back(res.has_length(count));
+      symbol_exprt idx=fresh_univ_index("QA_index_value_of");
+      equal_exprt eq(str[plus_exprt(idx, offset)], res[idx]);
+      string_constraintt a1(idx, count, eq);
+      axioms.push_back(a1);
+      return res;
+    }
+  else
+    {
+      assert(args.size()==1);
+      return add_axioms_for_java_char_array(args[0]);
+    }
+}
+
+string_exprt string_constraint_generatort::add_axioms_for_substring
+(const function_application_exprt &f)
+{
+  assert(f.arguments().size()>=2);
```

Comment:
You may want to introduce `const function_application_exprt::argumentst &args=f.arguments();` as you are making several references to `arguments()`